### PR TITLE
Default to one shard

### PR DESCRIPTION
**Pull request by jasontedor**
This commit changes the default out-of-the-box configuration for the number of shards from five to one. We think this will help address a common problem of oversharding. For users with time-based indices that need a different default, this can be managed with index templates. For users with non-time-based indices that find they need to re-shard with the split API in place they no longer need to resort only to reindexing.

Since this has the impact of changing the default number of shards used in REST tests, we want to ensure that we still have coverage for issues that could arise from multiple shards. As such, we randomize (rarely) the default number of shards in REST tests to two. This is managed via a global index template. However, some tests check the templates that are in the cluster state during the test. Since this template is randomly  there, we need a way for tests to skip adding the template used to set  the number of shards to two. For this we add the default_shards feature skip. To avoid having to write our docs in a complicated way because sometimes they might be behind one shard, and sometimes they might be behind two shards we apply the default_shards feature skip to all docs tests. That is, these tests will always run with the default number of shards (one).

### CODE REVIEW

**Comment by jasontedor** [comment]
FYI @elastic/es-clients for the feature skip `default_shards`. Note that you should not need to be concerned with this at all as the randomization to add a global template to set `index.number_of_shards` to 2 is done via our test runner so you will always pick up the new product default of 1. Thus, you do not need to skip tests that have this feature skip.

**Review by rjernst — APPROVED** [review]
LGTM, this was much simpler (fewer changes needed) than I would have anticipated!

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java [hunk 1]
```diff
@@ -367,7 +367,7 @@ public ClusterState execute(ClusterState currentState) throws Exception {
                 // now, put the request settings, so they override templates
                 indexSettingsBuilder.put(request.settings());
                 if (indexSettingsBuilder.get(SETTING_NUMBER_OF_SHARDS) == null) {
-                    indexSettingsBuilder.put(SETTING_NUMBER_OF_SHARDS, settings.getAsInt(SETTING_NUMBER_OF_SHARDS, 5));
+                    indexSettingsBuilder.put(SETTING_NUMBER_OF_SHARDS, settings.getAsInt(SETTING_NUMBER_OF_SHARDS, 1));
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java]: should this be 5 if the index.version.created is < 7.0 I mean we can still have a mixed cluster here no?

**Review by s1monw — COMMENTED** [review]
left one comment LGTM otherwise

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java (hunk 1 above)
Comment (jasontedor) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java]: @s1monw I pushed 2e89d980735b31bb73371eb5f8c37d2ab8c38a34.

**Comment by s1monw** [comment]
Ship it

**Comment by clintongormley** [comment]
Let's make sure we blog about this to explain the reasoning.

**Comment by jasontedor** [comment]
@robcowart I am so sorry for the slow reply here. Thanks so much for such a thoughtful question. Briefly, we think that the benefits of keeping shard counts as low as possible will benefit more users than shipping with a default configuration that will be beneficial to high-throughout use-cases. We aim to scalable in this regard, but we have a lot of users for which a single shard will suffice to absorb their write traffic. For users that need to scale, we have many knobs they can tune which include increasing their default shard count as well as splitting indices that were created with a single shard. We encourage such users to do performance testing to ensure that the tradeoff or doubling or more their shard count is worth the performance gain.

**Pull request merged**